### DISCUSSION TITLE
How can I deploy api and static site at the same time?

### DISCUSSION THREAD

**Discussion by gavinmcfarland** [vercel/vercel#4686] [Help] [answered]
Any help on the following issue would be greatly appreciated.

I have created a tool which lets me deploy a RESTlike API where I can fetch content.

I'm deploying it using the following configuration.

```json
{
	"version": 2,
	"public": true,
	"builds": [
		{
			"src": "api.js",
			"use": "@now/node-server",
			"config": {
				"includeFiles": [ "content/**" ]
			}
		}
	],
	"rewrites": [
		{ "source": "/api/(.*)", "destination": "/api.js" }
	]
}
```

This however, disables my static site from also deploying.

I'm using Sapper and have selected sapper as the framework from the Vercel dashboard.

What do I need to do in order to get Vercel to also deploy/reference the static site generated by `sapper export`.

**Comment by mcsdevv** [comment]
Hello,

I'd strongly recommend not trying to deploy an API that follows a legacy model of using a running server, especially inside a Serverless Function. It is not suited to the Vercel platform as is it serverless.

Instead, you can deploy your static frontend alongside [Serverless Functions](https://vercel.com/docs/v2/serverless-functions/introduction) in the `/api` directory (no configuration required).

**Reply by gavinmcfarland** [reply]
I believe this is not a viable solution for my needs as severless functions placed inside the `/api` folder do not easily give access to the URL and params for multiple path segments `/level1/level2/level3/` etc. I attempted to do this following the docs and found it very difficult to route my application effectively. Where as creating a custom application I'm able to manage the routing myself. Correct me if I've understood this wrong.

**Reply by gavinmcfarland** [reply]
Also I have a few questions. What is a legacy model of running a server? Did you mean to say you strongly recommend against doing that? Are there any more sophisticated examples or tutorials showing how to create a Node.js serverless function than the ones in the Vercel docs on serverless functions. Thanks.

**Reply by mcsdevv** [reply]
@limitlessloop you can use [path segments](https://vercel.com/docs/v2/serverless-functions/introduction#path-segments) to achieve this type of routing. However, the Vercel routing is file-based by default, relying on this approach where possible will make things easier.

By 'legacy model' I mean that you are trying to run a server inside of a Serverless Function, this is not recommended. The `@now/node-server` package is deprecated in favor of `@now/node` so please give that a try instead.

If you can let me know what you're trying to achieve, I can try and point you in the right direction for an example.

**Reply by gavinmcfarland** [reply]
Hi, thanks for your reply. The trouble with path segments in Vercel is they don't support params as you can in Express (from what I remember). Path segments if I understand correctly only give you one path segment. It's incredibly tedious to create something like this.

```js
app.get(base + ':resource1/:resource2', (req, res) => {
    getContent(database, {
        resource1: req.params.resource1,
        resource2: req.params.resource2,
        query: req.query
    }).then((value) => {
        if (value) {
            res.json(value);
        } else {
            res.send(`[Stancy] No value that matches query \n ${req.url}`);
        }
    });
});
```
I've created a tool which lets you create your own RESTlike API from static files. It generates an endpoint for each collection and item. Therefore it needs to know the name of each path segment and query (for example `?url=/page/about`) to be able to determine what content should be sent back. Is there an easier way to do this using what's available with Serverless Functions in Vercel?

Also, my tool reads content from a folder to create a JSON response. If using the `api/` directory how do you tell Vercel where the content is, similar to how you can with `builds.config.includeFiles`?

This is why it is much easier, if not the only way, to create my own app and to serve it using a custom configuration.

**Comment by gavinmcfarland** [comment, marked as answer]
Until a better answer can be given I'm going to describe it from my point of view.

When you use a `config` file with a `builds` property this overrides the default static build that Vercel performs.

Vercel's Serverless Functions are somewhat limited and uses `filesystem` routing. The documentation says it uses this by default, but it's not clear if this can be turned off per se. I think what the documentation is trying to suggest is simply that additional routing can be [configured](https://vercel.com/docs/configuration/#introduction/configuration-reference) differently if desired. What I can't tell is if routing can be configured to simply give a Serverless Function the whole URL. Perhaps this is possible but I cannot tell how from the documentation.

When using the `api/` folder for Serverless Functions there is no way to tell Vercel that there are files that the function may need to use during a build. Using a `config` file you can use the following.

```json
{
	"config": {
		"includeFiles": ["content/**"]
	}
}
```

For these reasons the only way to get around this is to serve the API with a `config` file using a legacy model of running server as @mcsdevv explained.

For the time being the way to achieve this concurrently would be to use a config file like the one below.

```json
{
	"version": 2,
	"public": true,
	"builds": [
		{
			"src": "__sapper__/export/**",
			"use": "@now/static",
			"config": {
				"includeFiles": ["content/**"]
			}
		},
		{
			"src": "api.js",
			"use": "@now/node",
			"config": {
				"includeFiles": ["content/**"]
			}
		}
	],
	"rewrites": [
		{ "source": "/api/(.*)", "destination": "/api.js" },
		{ "source": "/(.*)", "destination": "/__sapper__/export/$1" }
	]
}
```

Feel free to correct me if I have got any of this wrong.
